redash/models.py

```diff
@@ -757,6 +776,190 @@ def __unicode__(self):
+    object_type = peewee.CharField(index=True)
+    object_id = peewee.IntegerField(index=True)
+    object = GFKField('object_type', 'object_id')
+    access_type = peewee.CharField()
+    grantor = peewee.ForeignKeyField(User, related_name='grantor')
+    grantee = peewee.ForeignKeyField(User, related_name='grantee')
+
+    ACCESS_TYPE_VIEW = 'view'
+    ACCESS_TYPE_MODIFY = 'modify'
+    ACCESS_TYPE_DELETE = 'delete'
+
+    class Meta:
+        db_table = 'access_permissions'
+
+    @staticmethod
+    def grant_permission(object_type, object_id, access_type, grantee, grantor):
+        perm = AccessPermission()
+        perm.object_type = object_type
+        perm.object_id = object_id
+        perm.access_type = access_type
+        perm.grantor = grantor
+        perm.grantee = grantee
+        perm.save()
+
+    @staticmethod
+    def revoke_permission(object_type, object_id, grantee, access_type=None):
+        permissions = AccessPermission.find(object_id=object_id,
+            object_type=object_type, grantee=grantee, access_type=access_type)
+
+        if permissions.count() <= 0:
+            return None
+
+        permissions_deleted = list(permissions)
+
+        for permission in permissions:
+            permission.delete_instance()
+
+        if len(permissions_deleted) == 1:
+            return permissions_deleted[0]
+
+        return permissions_deleted
+
+    @staticmethod
+    def find(object_id, object_type, grantee=None, access_type=None, grantor=None):
+        q = AccessPermission.select(AccessPermission)\
+            .where(AccessPermission.object_id == object_id)\
+            .where(AccessPermission.object_type == object_type)
+        if grantee:
+            q = q.where(AccessPermission.grantee == grantee)
+        if access_type:
+            q = q.where(AccessPermission.access_type == access_type)
+        if grantor:
+            q = q.where(AccessPermission.grantor == grantor)
+        return q
+
+    @staticmethod
+    def exists(object_id, object_type, grantee, access_type=None):
+        try:
+            permissions = AccessPermission.find(object_id=object_id,
+                object_type=object_type, grantee=grantee, access_type=access_type)
+            return permissions.count() > 0
+        except Exception, e:
+            logging.warn("Unable to query for AccessPermission: %s" % e)
+            return False
+
+    @staticmethod
+    def reset(object_type, object_id, access_type):
+        revoke_permission(object_type=object_type, object_id=object_id, access_type=access_type)
+
+    def delete_instance(self, *args, **kwargs):
+        super(AccessPermission, self).delete_instance(*args, **kwargs)
+
+    def to_dict(self):
+        d = {
+            'id': self.id,
+            'object_id': self.object_id,
+            'object_type': self.object_type,
+            'access_type': self.access_type,
+            'grantor': self.grantor.id,
+            'grantee': self.grantee.id
```

Comment:
Use `self.grantor_id` and `self.grantee_id` or it will query the database for the user models here...